### PR TITLE
Java interview Марков

### PR DESCRIPTION
**Pull request by spirit93**

### CODE REVIEW

File: TaskManager.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    @Transactional
+    public void taskFinished(Task task) {
+        running.remove(task);
+        entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
+                " finished " + System.currentTimeMillis()));
+
+        blocked.stream().filter(task1 -> task1.getDependsOn().contains(task)).forEach(task1 -> {
```
Comment (spirit93) [inline TaskManager.java:46]: вынести в отдельную функцию

File: TaskManager.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            if (Collections.disjoint(task1.getDependsOn(), running)) {
+                blocked.remove(task1);
+                runTask(task1);
+            }
+        });
+    }
+
+    public void skeduleTask(Task task) {
```
Comment (HOOTORO) [inline TaskManager.java:54]: scheduleTask

File: TaskManager.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Transactional
+    public void taskFinished(Task task) {
+        running.remove(task);
+        entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
+                " finished " + System.currentTimeMillis()));
+
+        blocked.stream().filter(task1 -> task1.getDependsOn().contains(task)).forEach(task1 -> {
+            if (Collections.disjoint(task1.getDependsOn(), running)) {
```
Comment (HOOTORO) [inline TaskManager.java:47]: extract function isTaskBlocked

File: TaskManager.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                queue.add(i, task);
+                return;
+            }
+    }
+
+    public boolean isRunning(String name){
+        return running.stream().anyMatch(task -> task.getName() == name);
+    }
+
+    private void checkRun() {
```
Comment (HOOTORO) [inline TaskManager.java:66]: why checkRun? name doesn't corresponds what function actually do

File: TaskManager.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return instance;
+    }
+
+    @Transactional
+    public void runTask(Task task) {
+        if (Collections.disjoint(task.getDependsOn(), running)) {
+            running.add(task);
+            task.setStatus(TaskStatus.Working);
+            new Thread(() -> task.runTask(entityManager)).run();
+            entityManager.persist(new TaskHistoryRecord("task " + task.getName() + " " + task.getSchedule() + " " +
```
Comment (HOOTORO) [inline TaskManager.java:30]: use StringBuilder or String.format for better readability

File: TaskManager.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        checker.run();
+    }
+
+    public static TaskManager getInstance() {
+        return instance;
+    }
+
+    @Transactional
+    public void runTask(Task task) {
+        if (Collections.disjoint(task.getDependsOn(), running)) {
```
Comment (HOOTORO) [inline TaskManager.java:26]: anyMatch?

File: TaskManager.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            running.add(task);
```
Comment (HOOTORO) [inline TaskManager.java:27]: to remove duplicated code extract function addQueneTask(statusQuene HashSet<Task>, t Task)

File: TaskManager.java (hunk 5 above)
Comment (HOOTORO) [reply TaskManager.java:30]: or even better extract persisting to separate func

File: TaskHistoryRecord.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    private long id;
+    private String record;
+
+    public TaskHistoryRecord(String record) {
+        this.record = record;
+    }
+
+
+    public long getId() {
+        return id;
```
Comment (HOOTORO) [inline TaskHistoryRecord.java:14]: this.id

File: TaskHistoryRecord.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public String getRecord() {
+        return record;
```
Comment (HOOTORO) [inline TaskHistoryRecord.java:22]: this.record

File: Task.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    protected void reschedule(long timestamp){
+        schedule = timestamp;
+        TaskManager.getInstance().taskFinished(this);
+        TaskManager.getInstance().skeduleTask(this);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Task)) return false;
```
Comment (HOOTORO) [inline Task.java:23]: why not simplyfy to if instanceof && task.name?

File: Task.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    public abstract void runTask(EntityManager entityManager);
+
+    protected void finish(){
+        status = TaskStatus.Finished;
+        TaskManager.getInstance().taskFinished(this);
+    }
+
+    protected void reschedule(long timestamp){
+        schedule = timestamp;
+        TaskManager.getInstance().taskFinished(this);
```
Comment (HOOTORO) [inline Task.java:16]: unify method names 
finishTask, scheduleTask
or
taskScheduled, taskFinished

**Pull request closed without merging**